src/metatags/seo-checks.js

```diff
@@ -0,0 +1,209 @@
+      }
+      if (issue) {
+        this.detectedTags[url] ??= {};
+        this.detectedTags[url][tagName] ??= {};
+        Object.assign(this.detectedTags[url][tagName], {
+          [SEO_IMPACT]: issueImpact,
+          [ISSUE]: issue,
+          [ISSUE_DETAILS]: issueDetails,
+          [SEO_RECOMMENDATION]: recommendation,
+          ...(tagContent && { tagContent }),
+        });
+      }
+    };
+    checkTag(TITLE, pageTags[TITLE]);
+    checkTag(DESCRIPTION, pageTags[DESCRIPTION]);
+    checkTag(H1, (pageTags[H1] && pageTags[H1][0]) ? pageTags[H1][0] : null);
+  }
+
+  /**
+   * Checks if there are more than one H1 tags and adds to detected tags array if found lacking.
+   * @param {string} url - The URL of the page.
+   * @param {object} pageTags - An object containing the tags of the page.
+   */
+  checkForH1Count(url, pageTags) {
+    if (pageTags[H1]?.length > 1) {
+      this.detectedTags[url] ??= {};
+      this.detectedTags[url][H1] = {
+        tagContent: JSON.stringify(pageTags[H1]),
+        [SEO_IMPACT]: MODERATE,
+        [ISSUE]: MULTIPLE_H1_ON_PAGE,
+        [ISSUE_DETAILS]: `${pageTags[H1].length} H1 detected`,
+        [SEO_RECOMMENDATION]: ONE_H1_ON_A_PAGE,
+      };
+    }
+  }
+
+  /**
+   * Checks for tag uniqueness and adds to detected tags array if found lacking.
+   */
+  checkForUniqueness() {
+    [TITLE, DESCRIPTION, H1].forEach((tagName) => {
+      Object.values(this.allTags[tagName]).forEach((value) => {
+        if (value?.pageUrls?.size > 1) {
+          const capitalisedTagName = SeoChecks.#capitalizeFirstLetter(tagName);
+          const pageUrls = [...value.pageUrls];
+          pageUrls.forEach((url, index) => {
+            this.detectedTags[url] ??= {};
+            this.detectedTags[url][tagName] = {
+              tagContent: value.tagContent,
+              [SEO_IMPACT]: HIGH,
+              [ISSUE]: `Duplicate ${capitalisedTagName}`,
+              [ISSUE_DETAILS]: `${pageUrls.length} pages share same ${tagName}`,
+              [SEO_RECOMMENDATION]: UNIQUE_ACROSS_PAGES,
+              [DUPLICATES]: [
+                ...pageUrls.slice(0, index),
+                ...pageUrls.slice(index + 1),
+              ],
+            };
+          });
+        }
+      });
+    });
+  }
+
+  /**
+   * Adds tag data entry to all Tags Object
+   * @param url
+   * @param tagName
+   * @param tagContent
+   */
+  addToAllTags(url, tagName, tagContent) {
+    if (!tagContent) {
+      return;
+    }
+    const tagContentLowerCase = tagContent.toLowerCase();
+    this.allTags[tagName][tagContentLowerCase] ??= {
+      pageUrls: new Set(),
+      tagContent,
+    };
+    this.allTags[tagName][tagContentLowerCase].pageUrls.add(url);
+  }
+
+  /**
+   * Performs all SEO checks on the provided tags.
+   * @param {string} url - The URL of the page.
+   * @param {object} pageTags - An object containing the tags of the page.
+   */
+  performChecks(url, pageTags) {
+    this.checkForMissingTags(url, pageTags);
```

Comment:
add early return if
- url is not valid (use `isValidURL`)
- pageTags is not an object (use isObject or isNonEmptyObject)

possibly wrap those checks into class method

Reply:
The url here is actually the endpoint (path) of the complete url. I can use `hasText` to validate it.
Checking the pageTags using isObject.

Reply:
then it should be renamed to make this clear

Reply:
sure, renaming the url